Use Dims margin in layer slicing code

brisvag opened the pull request:
# Description
Followup to #5522 which uses the information about the "thick slices" (Dims `margin`s) in the slicing code.

All layers now have a `projection_mode` attribute, but only `Points`, `Vectors` and `Image` implement projections different from `none`.

Here's an example showcasing all the layers, with a homemade slider to control the `Dims.thickness`:

```py
import napari
import numpy as np
from skimage.data import binary_blobs
from scipy import ndimage as ndi
from magicgui import magicgui
from vispy.geometry import create_sphere

@magicgui(
    auto_call=True,
    thickness=dict(widget_type='FloatSlider', min=0, max=50)
)
def change_thickness(viewer: napari.Viewer, thickness=10):
    v.dims.thickness = (thickness,) * v.dims.ndim

v = napari.Viewer()
v.window.add_dock_widget(change_thickness)

img = binary_blobs(100, volume_fraction=0.1, n_dim=4)
pts = np.random.rand(100, 4) * 100
vecs = np.stack([pts, np.random.rand(100, 4) * 20 - 10], axis=1)
shapes = np.random.rand(4, 2) * 100
labels = ndi.label(img)[0]
sphere = create_sphere(method='ico')
surface = sphere.get_vertices() * 50 + 50, sphere.get_faces()
tracks = np.random.rand(50, 5) * 100
tracks[:, 0] = np.repeat(np.arange(10), 5)

il = v.add_image([img, img[::2, ::2, ::2, ::2]], projection_mode='mean')
pl = v.add_points(pts, projection_mode='all')
vl = v.add_vectors(vecs, projection_mode='all')
shl = v.add_shapes(shapes)
ll = v.add_labels(labels)
sul = v.add_surface(surface)
tl = v.add_tracks(tracks)

change_thickness()
```

https://github.com/napari/napari/assets/23482191/6b2b5524-90a9-4e6e-a388-f2aa79a0f8a5


# References
- First attempt at implementation with discussion and references: #5522
- closes #3600 (yay!)
- supersedes #1820
- related #1744

brisvag committed:
somewhat works

brisvag committed:
fix wrong loops

brisvag committed:
readd backward compatible current_step

brisvag committed:
forgotten line

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice2

brisvag committed:
fix tests

brisvag committed:
some docstring and typing updates + new useful setters

brisvag committed:
prevent set_thickness and set_point to go beyong range limits + fix tests

brisvag committed:
round used to no return int

brisvag committed:
switch to evented models

brisvag committed:
converting to lists...

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice2

brisvag committed:
merge-squash feature/evented-properties2

brisvag committed:
Merge branch 'feature/thick-slice2' into feature/dims-span

brisvag committed:
continue merging old pr code

brisvag committed:
moving along

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
remove useless property view

brisvag committed:
fix separation for range/span in viewermodel

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
Force span and step rather than erroring out

brisvag committed:
fix dims tests

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
pass some tests

brisvag committed:
fix private call and test

brisvag committed:
switch remaining current_step to point_step

brisvag committed:
Merge branch 'main' into feature/dims-span

brisvag committed:
revert leftover from old changes

brisvag committed:
fix dependend props calculation

brisvag committed:
recursively check for dependents

brisvag committed:
add smart event firing for properties

brisvag committed:
Merge branch 'fix/evented-dependent-properties' into feature/dims-span

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
revert leftover from old changes

brisvag committed:
fix remaining tests

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
fix midpoint calculation (use slider coords)

brisvag committed:
Merge branch 'main' into feature/dims-span

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
Merge branch 'feature/dims-span' of github.com:brisvag/napari into feature/dims-span

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/dims-span

brisvag committed:
Merge branch 'feature/dims-span' of github.com:brisvag/napari into feature/dims-span

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
switch to point/margins system and bring back 3-tuple range but with named tuple

brisvag committed:
fix namedtuple annotation

brisvag committed:
fix qt dims test

brisvag committed:
undo wrong change

brisvag committed:
fix more tests

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

jni committed:
Merge branch 'main' into feature/dims-span

brisvag committed:
address comments

brisvag committed:
revert name change

brisvag committed:
Update napari/components/dims.py

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

brisvag committed:
fix order check, add translations and remove old docstring

brisvag committed:
do not clip step to max (would fail with degenerate dimensions)

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/dims-span

brisvag committed:
remove leftover setter to minimize changes

brisvag committed:
fix var names

brisvag committed:
some stuff works

brisvag committed:
fix points thick slice

brisvag committed:
add some comments

brisvag committed:
implement slicing for (single scale) images

brisvag committed:
add some tests, address review comments

brisvag committed:
revert useless test changes

brisvag committed:
fix test

brisvag committed:
Empty commit

andy-sweet commented on napari/layers/utils/_slice_input.py:
```diff
@@ -79,14 +81,31 @@ def data_indices(
+            data_margin_right = np.round(data_margin_right).astype(int)
 
-        indices = [slice(None)] * self.ndim
+        indices = [[np.nan, np.nan, np.nan] for _ in range(self.ndim)]
```
Does this make the return type of this method `Tuple[Tuple[Union[float, int], Union[float, int], Union[float, int]]]` now?

brisvag replied:
Yes, this needs to be changed, and it's in fact one of the things where we might need to change a bit more deeply the code. Thing is, the return of this method used to be basically plug and play for slicing; now, however, it's a completely different 3-tuple with point, margin left and margin right.

It might be worth renaming this method as well.

Maybe indeed having a dataclass for this little piece of information is a good idea...

brisvag committed:
Merge branch 'main' into feature/thick-slice-layers

brisvag committed:
encapsulate thick slice dataclass

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
request/response for vectors layer

melonora committed:
Merge branch 'main' into feature/vector-response

brisvag committed:
happy mypy

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/vector-response

brisvag committed:
Merge branch 'feature/vector-response' of github.com:brisvag/napari into feature/vector-response

brisvag committed:
Merge branch 'feature/vector-response' into feature/thick-slice-layers

brisvag committed:
thick slicing for vectors

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
kinda working with all other layers (labels broken)

Czaki committed:
Pin numpy below 2 (#6073)

# Description
As advertised at scipy 2023 we should pin numpy below 2

## Type of change
- [x] Maintenance (changes required to run napari, tests, & CI smoothly)

dstansby committed:
Fix some typing in napari._vispy (#6063)

# Description
Fixed some typing in `napari._vispy` modules.

# References
More work towards #2751

andy-sweet commented on napari/layers/base/base.py:
```diff
@@ -904,12 +908,9 @@ def _clear_extents_and_refresh(self):
+    def _data_slice(self) -> _ThickNDSlice:
+        """Slice in data coordinates."""
+        return self._slice_input.data_slice(
             self._data_to_world.inverse,
```
One problem here is that `self._data_to_world.inverse` is non-trivial to evaluate. For async slicing, I allow it because in that case there are typically more pressing performance issues.

But for the default sync-slicing in cases when no slicing actually needs to be done, this could be a more important cost. I (and probably others) have thought about caching `_data_to_world` and its `inverse` in some way, which should make this a non-issue. Or maybe we can delay its evaluation in `_SliceInput.data_slice` and return early in cases where it's not needed?

Otherwise, could we early return with a "full" slice here?

brisvag replied:
I reimplemented the early return... Definitely should figure out the rest as well, but at least this shouldn't be a regression now.

andy-sweet commented on napari/layers/vectors/vectors.py:
```diff
@@ -679,17 +683,17 @@ def _make_slice_request(self, dims) -> _VectorSliceRequest:
         # or by lazily evaluating it in the slice task itself.
         slice_indices = slice_input.data_indices(
-            self._data_to_world.inverse, round_index=False
+            self._data_to_world.inverse, round_index=True
```
You're changing `round_index` from `False` to `True` for points and vectors here, but I'm not sure why. At the current commit, I think `round_index` is always `True`, so if you can justify the change, we could just delete it entirely.

brisvag replied:
I think this could be a leftover from refactors... I remember we decussed this line before elsewhere; was the conclusion that points and vectors should both have no rounding?

andy-sweet replied:
Yes, that's what I remember.

andy-sweet commented on napari/layers/vectors/_slice.py:
```diff
@@ -47,15 +47,15 @@ class _VectorSliceRequest:
         The layer's data field, which is the main input to slicing.
-    dims_indices : tuple of ints or slices
+    data_slice : tuple of ints or slices
         The slice indices in the layer's data space.
```
name is updated, but type and description need to be updated too

andy-sweet commented on napari/layers/image/_slice.py:
```diff
@@ -264,6 +264,27 @@ def _slice_indices_at_level(self, level: int) -> np.ndarray:
+            slices.append(slice_)
+
+        # TODO: for now just average is implemented over thick slice
+        return np.mean(
```
Just a reminder, but this might explain the dtype issue with labels because slicing labels uses the image slice code.

brisvag replied:
This code is quite different now: labels don't have this issue anymore cause they only project in `none` mode :)

andy-sweet commented on napari/layers/utils/_slice_input.py:
```diff
@@ -78,17 +106,41 @@ def data_indices(
+        return _ThickNDSlice(
+            point=full_pts,
+            margin_left=full_margin_left,
+            margin_right=full_margin_right,
```
Based on the typing of `_ThickNDSlice`, I think this needs to be:

```suggestion
            point=tuple(full_pts),
            margin_left=tuple(full_margin_left),
            margin_right=tuple(full_margin_right),
```

Changing the typing in the class is also fine with me, but i do love immutable things :)

andy-sweet commented on napari/layers/utils/_slice_input.py:
```diff
@@ -78,17 +106,41 @@ def data_indices(
-        indices = [slice(None)] * self.ndim
-        for i, ax in enumerate(self.not_displayed):
-            indices[ax] = data_pts[i]
+        full_pts = [np.nan for ax in range(self.ndim)]
```
```suggestion
        full_pts = [np.nan for _ in range(self.ndim)]
```

andy-sweet commented on napari/layers/utils/_slice_input.py:
```diff
@@ -78,17 +106,41 @@ def data_indices(
+
         data_pts = slice_world_to_data(world_pts)
+        data_margin_left = slice_world_to_data(world_margin_left)
+        data_margin_right = slice_world_to_data(world_margin_right)
```
This seems like a slow way to apply the same transform to 3 coordinates. Is there some way to stack these and just apply it once?

andy-sweet reviewed (changes requested):
I'm good with the overall approach here, but there are still a few important details to sort out. Let me know if you want me to try solving some of those.

brisvag committed:
projection works

brisvag committed:
some changes from review

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
address some comments

brisvag committed:
fix wrong unstack

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
fixes and simplification by adding some utility methods to _ThickNDSlice

brisvag committed:
fix image slice

brisvag committed:
fix viewer and slicing

brisvag committed:
fix flipped

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
fix mistakenly changed code

brisvag committed:
fix sign

brisvag committed:
correctly init proj mode

brisvag committed:
fix tests

brisvag committed:
fix some tests

andy-sweet commented on napari/layers/image/_slice.py:
```diff
@@ -255,14 +262,46 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
+        if self.projection_mode == 'none':
+            # early return with only the dims point being used
+            indices = np.array(slice_not_disp.point, dtype=int)
+            return data[tuple(indices)]
```
If the data being sliced causes lazy evaluation when indexed (e.g. a dask array), then we need `asarray` (or something similar) to force it to be materialized/computed.

```suggestion
            # Ensure any lazy arrays are materialized.
            return np.asarray(data[tuple(indices)])
```

This should fix the dask test failures. Like I said, they're sometimes annoying, but sometimes make good points/catches too :)

brisvag replied:
same thing in the return below that I suppose :+1:

andy-sweet replied:
Right now, it's may not be necessary because I think the calls to `np.sum` and `np.mean` should also force evaluation, but I guess it won't do much harm either.

brisvag committed:
materialize dask arrays to reduce fetches

brisvag committed:
add kwargs everywhere

brisvag committed:
fix surface slicing

brisvag committed:
pass proj modes throu str() to ensure conversion works (label<->image)

brisvag committed:
fix missing slicing

brisvag committed:
coerce proj mode to str so it works across enums

brisvag committed:
extract some more methods to thickslice and fix several tests

brisvag committed:
some more test fixes

brisvag committed:
fix wrong indexing for no projection image

brisvag committed:
Merge branch 'feature/thick-slice-layers' of github.com:brisvag/napari into feature/thick-slice-layers

brisvag committed:
fix more tests and slicing order

brisvag committed:
restore non-rounded indexing

brisvag committed:
get rid of round index

brisvag committed:
fix image rounding

brisvag committed:
ensure slice input always has non-nans in nondisplayed dims

brisvag committed:
fix rounding for surf

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
happy mypy?

brisvag committed:
fix image slicing

brisvag committed:
add some projections for image and add test

brisvag committed:
add simple gui

andy-sweet commented on napari/utils/misc.py:
```diff
@@ -148,8 +148,9 @@ def is_iterable(arg, color=False, allow_none=False):
+        isinstance(arg, (str, Enum))
+        or arg is None
+        and not allow_none
         or np.isscalar(arg)
```
Nit: this at least needs some parentheses as before. But given the increased complexity, maybe can we just break this into a couple of early return conditions? I think that helps in this particular case, so this effectively ends up like a switch statement.

andy-sweet commented on napari/layers/vectors/_slice.py:
```diff
@@ -81,21 +83,7 @@ def __call__(self) -> _VectorSliceResponse:
                 request_id=self.id,
             )
 
-        # We want a numpy array so we can use fancy indexing with the non-displayed
```
Do you have sense or evidence about performance changes here when not doing thick-slicing? Could we do a before/after run on one of the existing ASV benchmarks (e.g. `benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice`).

Similar question for the other potentially affected layers here - I think that's images and points?

brisvag replied:
I tried a few menchmarks, and it seems that we have a regression, reproducible over a couple runs. TLDR: The regression seems bigger for small data, which is at least promising.

Here's vectors:

```
[ 50.00%] · For napari commit e6815274 <feature/thick-slice-layers> (round 2/2):
[ 50.00%] ·· Benchmarking virtualenv-py3.11
[ 75.00%] ··· benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice                                                                                                                                                              ok
[ 75.00%] ··· ======== =============
               param1
              -------- -------------
                 16     1.28±0.02ms
                 64     1.27±0.02ms
                256     1.35±0.02ms
                1024    1.50±0.01ms
                4096    2.06±0.08ms
               16384    4.16±0.06ms
               65536     12.8±0.2ms
              ======== =============

[ 75.00%] · For napari commit 21bc6244 <main> (round 2/2):
[ 75.00%] ·· Building for virtualenv-py3.11.........................
[ 75.00%] ·· Benchmarking virtualenv-py3.11
[100.00%] ··· benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice                                                                                                                                                              ok
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 16       993±30μs
                 64     1.00±0.03ms
                256     1.09±0.05ms
                1024    1.29±0.06ms
                4096    1.83±0.09ms
               16384    3.83±0.08ms
               65536     12.2±0.2ms
              ======== =============

       before           after         ratio
     [21bc6244]       [e6815274]
     <main>           <feature/thick-slice-layers>
+        993±30μs      1.28±0.02ms     1.29  benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice(16)
+     1.00±0.03ms      1.27±0.02ms     1.27  benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice(64)
+     1.09±0.05ms      1.35±0.02ms     1.23  benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice(256)
+     1.83±0.09ms      2.06±0.08ms     1.12  benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice(4096)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

And here's points:

```
[ 50.00%] · For napari commit e6815274 <feature/thick-slice-layers> (round 2/2):
[ 50.00%] ·· Benchmarking virtualenv-py3.11
[ 75.00%] ··· benchmark_points_layer.Points3DSuite.time_set_view_slice                                            ok
[ 75.00%] ··· ======== =============
               param1
              -------- -------------
                 16     1.35±0.03ms
                 64     1.40±0.05ms
                256     1.42±0.04ms
                1024    1.52±0.05ms
                4096    1.56±0.04ms
               16384     1.88±0.2ms
               65536    3.25±0.05ms
              ======== =============

[ 75.00%] · For napari commit 21bc6244 <main> (round 2/2):
[ 75.00%] ·· Building for virtualenv-py3.11.........................
[ 75.00%] ·· Benchmarking virtualenv-py3.11
[100.00%] ··· benchmark_points_layer.Points3DSuite.time_set_view_slice                                            ok
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 16     1.09±0.02ms
                 64     1.09±0.02ms
                256     1.15±0.07ms
                1024    1.13±0.01ms
                4096    1.22±0.03ms
               16384    1.57±0.05ms
               65536     3.03±0.4ms
              ======== =============

       before           after         ratio
     [21bc6244]       [e6815274]
     <main>           <feature/thick-slice-layers>
+     1.13±0.01ms      1.52±0.05ms     1.35  benchmark_points_layer.Points3DSuite.time_set_view_slice(1024)
+     1.09±0.02ms      1.40±0.05ms     1.29  benchmark_points_layer.Points3DSuite.time_set_view_slice(64)
+     1.22±0.03ms      1.56±0.04ms     1.28  benchmark_points_layer.Points3DSuite.time_set_view_slice(4096)
+     1.15±0.07ms      1.42±0.04ms     1.23  benchmark_points_layer.Points3DSuite.time_set_view_slice(256)
+     1.09±0.02ms      1.35±0.03ms     1.23  benchmark_points_layer.Points3DSuite.time_set_view_slice(16)
+     1.57±0.05ms       1.88±0.2ms     1.20  benchmark_points_layer.Points3DSuite.time_set_view_slice(16384)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

andy-sweet replied:
I'm not too troubled by the increases here because like you said, with larger numbers of elements, other things start to dominate. So it would be nice to minimize them, but if that's too difficult, I don't think they should block this PR.

What about images?

brisvag replied:
Also similar to the above (failed ones are because my laptop doens't have enough memory):

```
[ 50.00%] ··· Running (benchmark_image_layer.Image3DSuite.time_set_view_slice--).
[ 50.00%] · For napari commit e6815274 <feature/thick-slice-layers~1> (round 2/2):
[ 50.00%] ·· Benchmarking virtualenv-py3.11
[ 75.00%] ··· benchmark_image_layer.Image3DSuite.time_set_view_slice                                                                                                                                                          1/7 failed
[ 75.00%] ··· ======== =============
               param1
              -------- -------------
                 16     1.13±0.02ms
                 32     1.13±0.01ms
                 64     1.14±0.01ms
                128     1.13±0.03ms
                256     1.16±0.03ms
                512     1.20±0.04ms
                1024       failed
              ======== =============

[ 75.00%] · For napari commit b5ecd293 <main> (round 2/2):
[ 75.00%] ·· Building for virtualenv-py3.11..............
[ 75.00%] ·· Benchmarking virtualenv-py3.11
[100.00%] ··· benchmark_image_layer.Image3DSuite.time_set_view_slice                                                                                                                                                          1/7 failed
[100.00%] ··· ======== =============
               param1
              -------- -------------
                 16       969±7μs
                 32       960±9μs
                 64       966±10μs
                128       975±10μs
                256     1.03±0.01ms
                512     1.02±0.02ms
                1024       failed
              ======== =============

       before           after         ratio
     [b5ecd293]       [e6815274]
     <main>           <feature/thick-slice-layers~1>
+        966±10μs      1.14±0.01ms     1.18  benchmark_image_layer.Image3DSuite.time_set_view_slice(64)
+         960±9μs      1.13±0.01ms     1.18  benchmark_image_layer.Image3DSuite.time_set_view_slice(32)
+     1.02±0.02ms      1.20±0.04ms     1.18  benchmark_image_layer.Image3DSuite.time_set_view_slice(512)
+         969±7μs      1.13±0.02ms     1.17  benchmark_image_layer.Image3DSuite.time_set_view_slice(16)
+        975±10μs      1.13±0.03ms     1.16  benchmark_image_layer.Image3DSuite.time_set_view_slice(128)
+     1.03±0.01ms      1.16±0.03ms     1.13  benchmark_image_layer.Image3DSuite.time_set_view_slice(256)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

andy-sweet replied:
Similarly, I don't think the performance decrease blocks this PR. Sharing the same code path has its benefits and we could always follow up with some future optimizations.

andy-sweet commented on napari/layers/base/base.py:
```diff
@@ -1136,39 +1159,52 @@ def _slice_dims(
             ndisplay=ndisplay,
-            point=point,
-            order=order,
+            world_slice=world_slice[-self.ndim :],
```
Does `_ThickNDSlice.make_full` handle the world-to-layer dimensional correspondence that `_world_to_layer_dims` handles? Does it do in a different way? We already have a few functions that do this slightly differently (maybe for good reasons), but it would be nice to avoid different definitions and code/logic reuse if not needed.

brisvag replied:
`make_full` doesn't do any coordinate space transformation. It's staying in whatever coordinate space the inputs are. In fact, it's used both as a world space slice in `_SliceInput` and as a data space slice in the `SliceRequest` classes.

andy-sweet replied:
Oh I see, `order` is still used to handle that transformation. Thanks for clarifying!

andy-sweet commented on napari/layers/image/_tests/test_image.py:
```diff
@@ -874,3 +874,52 @@ def test_rendering_init():
     assert layer.rendering == ImageRendering.ISO.value
+
+
+def test_thick_slice():
```
Would be great to have one test case for multi-scale.

andy-sweet reviewed (changes requested):
Pretty close to approving this. Mostly need to see a few more tests and benchmarks. Though I also had some questions to clarify too.

brisvag committed:
some fixes and added test

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
revert qt control changes

brisvag committed:
add point test

andy-sweet commented on napari/components/viewer_model.py:
```diff
@@ -257,6 +257,8 @@ def __init__(
         self.dims.events.order.connect(self._update_layers)
         self.dims.events.order.connect(self.reset_view)
         self.dims.events.current_step.connect(self._update_layers)
```
Given that `point` is more like `margin_left` and `margin_right` (i.e. they are in world coordinates), I wonder if we should connect to that instead. I'm trying to think about the consequences of this, but we can almost certainly do it outside of this PR anyway.

brisvag replied:
Actually, I think we really should do it here, you're right. Right now, if you change `point` by a small enough amount that `current_step` is unaffected, nothing changes, and you end up in an inconsistent state.

andy-sweet replied:
Ah, I see. The margins in `Dims` are positive offsets from `point`? I think I forgot that. So moving `point` has the effect of moving the region we're averaging over? If so, then agreed the change should probably happen here.

brisvag replied:
Yep, marging left is subtracted from `point` and the right is added.

andy-sweet commented on napari/layers/utils/_slice_input.py:
```diff
@@ -47,48 +150,51 @@ def not_displayed(self) -> List[int]:
-        world_pts = [self.point[ax] for ax in self.not_displayed]
-        data_pts = slice_world_to_data(world_pts)
-        if round_index:
-            # A round is taken to convert these values to slicing integers
```
Is this a change in behavior for non-thick slicing of image layers? Or put another way, what happens when I have two image layers with different extents (either through shape or scale) then move the slider?

Or is this covered by the new logic in `_project_thick_slice`?

Sorry if I asked this before, but I think this is one of the potentially larger changes in this PR.

brisvag replied:
Not sure I get the question; you mean the fact that `np.round` is no longer here?
This is now happening in `_project_thick_slice`, because we need to keep full float precision until the last moment. In that method, after adding `margins` to the `point` to get the actual begin and end of the thick slice, we round the result before using it to index the layer data.

Not sure how this relates to extents, so maybe I misunderstood the question ^^'

andy-sweet replied:
I think the rounding only becomes interesting when you have two layers with different scales/shapes (which effectively determines their extents) - then the slider positions may not exactly hit the data indices of either layer and you need the rounding for things like images.

But thanks for confirming that this is happening in `_project_thick_slice`!

brisvag replied:
Ok! Btw, that used to be the only case, but now `Dims` can actually have floating point `point`, so this case can happen any time even with a single layer.

andy-sweet reviewed (commented):
Only remaining question is about the recent changes to rounding. I might be able to contribute some tests to cover this and understand the current/previous behavior, but probably not today.

Once that question is resolved, I'm approving this. Since there are significant changes here, you may want a second opinion, but I trust you to make that call. Thanks for all this work and patience!

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
connect to the right event

Czaki commented on napari/components/_layer_slicer.py:
```diff
@@ -226,7 +226,12 @@ def submit(
+                dims.margin_right,
+                dims.ndisplay,
+                dims.order,
+                force=force,
```
Why we do not pass dims object here?

brisvag replied:
Good question, it would be less noisy... I was following the existing pattern, so maybe @andy-sweet had a good reason for this?

andy-sweet replied:
Not really - I was following the pattern before me :D

In the long term, I want slicing to be functionally defined as `layer(dims) -> slice`, so I'm in favor of this change. Don't feel strongly about this happening in this PR though, especially as I'm unconvinced about the long term future of `Layer._slice_dims`. I think you'd mostly need to change a bunch of tests.

Czaki commented on napari/components/viewer_model.py:
```diff
@@ -726,6 +728,7 @@ def add_image(
         plane=None,
         experimental_clipping_planes=None,
         custom_interpolation_kernel_2d=None,
+        projection_mode='none',
```
We really need to reduce the number of arguments...

kevinyamauchi replied:
I agree. Let's address it in a separate issue/PR though.

Czaki commented on napari/layers/base/base.py:
```diff
@@ -1137,39 +1160,52 @@ def _slice_dims(
+        self,
+        point=None,
+        margin_left=None,
+        margin_right=None,
```
I would prefer to pass here only single argument (for example `margin`)

Something like:

```python
@dataclass
class Margin:
    left: Optional[Tuple]
    right: Optional[Tuple]
```

brisvag replied:
@jni, @andy-sweet, we touched on this before, but maybe not exactly for this context. What do you think? My current thinking is that having a class (or namedtuple) here might look nicer in the signatures and function calls, but it's less intuitive to work with (especially if users end up having to use it).

andy-sweet replied:
Isn't `Margin` almost `_ThickNDSlice`? Could we just create a `_ThickNDSlice` where it's called and pass that through instead of `point`, `margin_left`, and `margin_right`?

brisvag replied:
I tried doing that, but as suggested in an above comment by @Czaki it quickly just seemed better to pass `Dims` itself. The one thing I'm unsure about is whether this is risky due to threading, so maybe we need to make a copy of it?

andy-sweet replied:
Both `_make_slice_input` and `_slice_dims` should be executed on the main thread, so if passing through `Dims` is preferred, I think you should go for it.

Czaki commented on napari/layers/image/_image_utils.py:
```diff
@@ -105,3 +106,17 @@ def guess_labels(data):
     return 'image'
+
+
+def project_slice(data, axis, mode):
```
Why not `NONE` mode?

Will `np.sum` etc correctly work with Zarr, Dask, etc?

brisvag replied:
There's an early return in `_ImageSliceRequest._project_thick_slice`. This prevents for example dask array being sliced unnecessarily.

brisvag replied:
Yes, those np functions should work directly with zarr and dask, as far as I know.

Czaki commented on napari/layers/image/_slice.py:
```diff
@@ -255,14 +264,53 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
+
+        if self.projection_mode == 'none':
+            # early return with only the dims point being used
+            return np.asarray(data[tuple(slices)])
```
why we need to make a copy of data here (potential memory issue)

brisvag replied:
We added this here after debugging with @andy-sweet; the lazy arrays need to be computed here because otherwise there's several points later that re-slice them and our dask tests were failing.

Maybe this should actually be `dask.compute`, now that I think about it... @andy-sweet, was there another reason for this `asarray` other than materializing lazy arrays?

andy-sweet replied:
I think the main reason for `asarray` here is because `Image.data` only needs to implement the layer data protocol whereas vispy wants a numpy array. In the case that `data` needs to be materialized/evaluated (e.g. a dask array), then I think we want to do it here because at least we have a chance of doing it off the main thread (e.g. when async slicing is enabled).

We could special case for dask, but I don't see the benefit because the [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html#numpy-asarray) claim that `asarray` only makes a copy if it needs to and in this case it shouldn't if `data` is a numpy array because we're not changing the dtype or order. Is that understanding wrong? Should we use `np.array(..., copy=False)` instead?

The only other potential copy I can see is indexing `data` (because of [numpy advanced indexing](https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing)) , but I think the index is always a `tuple[int | slice, ...]` so that shouldn't happen.

Czaki commented on napari/layers/image/_slice.py:
```diff
@@ -255,14 +264,53 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
+        slice_arr[0] = np.clip(slice_arr[0], 0, self.level_shapes[level] - 1)
+        return _ThickNDSlice.from_array(slice_arr)
+
+    def _project_thick_slice(self, data, data_slice):
```
Please add type annotation. It is hard to read this function without knowing the type of data.

Czaki commented on napari/layers/utils/_slice_input.py:
```diff
@@ -2,14 +2,117 @@
+
+    def __iter__(self):
+        yield from zip(self.point, self.margin_left, self.margin_right)
+
```
It will be nice to document what is yield. After reading the whole class it is clear, but it is time consuming.

Czaki commented on napari/layers/base/base.py:
```diff
@@ -1117,7 +1134,13 @@ def _set_view_slice(self):
+        self,
+        point=None,
+        margin_left=None,
+        margin_right=None,
```
Merge these parameters into one and add to docstring?

Czaki commented on napari/layers/image/image.py:
```diff
@@ -735,26 +741,30 @@ def _set_view_slice(self) -> None:
+            dims.margin_left,
+            dims.margin_right,
+            dims.ndisplay,
+            dims.order,
```
Maybe just pass dims?

Czaki commented on napari/layers/points/_slice.py:
```diff
@@ -4,7 +4,8 @@
 
 from napari.layers.base._slice import _next_request_id
-from napari.layers.utils._slice_input import _SliceInput
+from napari.layers.points._points_constants import ProjectionMode
```
The longer I read this code the more I would like to have `PointsProjectionMode` to avoid confusion and faster code navigation

brisvag replied:
I could rename all these to match the layer class, you're right. Probably could do the same with layer `Mode`s I suppose!

Czaki commented on napari/layers/points/_tests/test_points.py:
```diff
@@ -2348,7 +2353,7 @@ def test_shown_view_size_and_view_data_have_the_same_dimension():
     assert layer._view_size.shape[0] == layer._view_data.shape[0]
     assert layer._view_size.shape[0] == 2
-    assert np.array_equal(layer._view_size, [3, 1])
+    assert np.array_equal(layer._view_size, [3, 2])
```
maybe? 
```suggestion
    assert np.array_equiv(layer._view_size, [3, 2])
```

Czaki commented on napari/utils/misc.py:
```diff
@@ -147,11 +147,11 @@ def is_iterable(arg, color=False, allow_none=False):
-        or isinstance(arg, str)
-        or np.isscalar(arg)
-    ):
+    if arg is None and not allow_none:
+        return False
+    if isinstance(arg, (str, Enum)):
+        return False
+    if np.isscalar(arg):
```
Why split on multiple clauses?

brisvag replied:
I thought it looked hard to parse when it was in a single clause. This is clearer because it's divided logically.

Czaki reviewed (commented):
Why do you choose `np.nan` instead `slice(None)`. I can't found this?

I know it is internal machinery, but I want to see more docstrings. Currently, it is hard to follow and looks hard to maintain. 

I know this problem does not come from this PR but from the general over growth of Layers.

brisvag committed:
add some typing

brisvag committed:
Merge branch 'feature/thick-slice-layers' of github.com:brisvag/napari into feature/thick-slice-layers

brisvag committed:
Update napari/layers/points/_tests/test_points.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

brisvag committed:
rename projection modes

brisvag committed:
use dims instead of passing individual parameters

brisvag committed:
hide behind type checking

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
hide more stuff behind TYPE_CHECKING

andy-sweet reviewed (approved):
I still need to pull down the latest version of this and test it, but I'm approving now since most of my concerns have been addressed and I don't see anything blocking. Given the size of the PR and some of the recent changes, it's hard to be certain, but I'm OK with a little uncertainty here given the utility of the new feature.

brisvag committed:
Merge branch 'main' of https://github.com/napari/napari into feature/thick-slice-layers

brisvag committed:
forgot to rename one thing

brisvag committed:
add some docstrings

kevinyamauchi commented on napari/layers/image/_slice.py:
```diff
@@ -251,22 +261,63 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
+            slice(None) if np.isnan(p) else int(np.round(p))
+            for p in data_slice.point
+        ]
+
```
It would be nice to have an explanatory comment

brisvag replied:
Added some comments.

kevinyamauchi commented on napari/layers/image/_slice.py:
```diff
@@ -251,22 +261,63 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
+            low = int(np.round(point - m_left))
+            high = int(np.round(point + m_right))
+
+            # if high is already at an integer, we need to round up to next intereger
```
```suggestion
            # if high is already at an integer, we need to round up to next integer
```

kevinyamauchi commented on napari/layers/image/_slice.py:
```diff
@@ -251,22 +261,63 @@ def _call_multi_scale(self) -> _ImageSliceResponse:
+        slices: List[slice | int] = [
+            slice(None) if np.isnan(p) else int(np.round(p))
+            for p in data_slice.point
+        ]
```
From our discussion, it seems like the slice rounding code might not quite cover all cases. I think we should pull out the slice rounding code into a util function and then make a set of tests that covers all cases we can think of.

brisvag replied:
Moved to separate methods, added comments, and added a test for it.

brisvag commented on napari/layers/image/image.py:
```diff
@@ -713,11 +716,14 @@ def _set_view_slice(self) -> None:
+            low = pt - data_slice.margin_left[d]
+            high = pt + data_slice.margin_right[d]
+            if pt < max(0, low) or pt > min(self._extent_data[1][d], high):
                 self._slice = _ImageSliceResponse.make_empty(
```
double check this code (should allow point to be outside of data if margin gets back in it)

brisvag replied:
Fixed this logic and moved it to the `_ImageSliceRequest` class

kevinyamauchi commented on napari/layers/utils/_slice_input.py:
```diff
@@ -2,14 +2,129 @@
+            (0,) * ndim if margin_right is None else tuple(margin_right)
+        )
+
+        prepend = max(ndim - val_ndim, 0)
+
+        point = (0,) * prepend + point
+        margin_left = (0,) * prepend + margin_left
+        margin_right = (0,) * prepend + margin_right
```
It would be nice to add a comment explaining this

brisvag replied:
added comments and streamlined this a bit.

kevinyamauchi commented on napari/layers/utils/_slice_input.py:
```diff
@@ -2,14 +2,129 @@
+
+        If ndim is provided, it will be used to crop or extend the given values.
+        Values not provided will be filled with np.nan for point and zeros for margins.
+        Prepended values are always zeros.
```
I find this function to be a bit complex because it does different things depending on the input. Would it be possible to break it up into a few functions, each for one type of input?

kevinyamauchi replied:
Also, can you please add some tests? I think this will be helpful both to document the intent of the function and for future maintenance.

kevinyamauchi commented on napari/layers/base/base.py:
```diff
@@ -1117,59 +1152,60 @@ def _set_view_slice(self):
-
-        if order is None:
-            order = tuple(range(ndim))
+        # if not point is given, "world" has same dimensionality of self
```
```suggestion
        # if no point is given, "world" has same dimensionality of self
```

kevinyamauchi reviewed (commented):
Thanks @brisvag ! I think this looks good. I've left a few suggestions and comments below

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
better check for out of bounds slice

brisvag committed:
Apply suggestions from code review

Co-authored-by: Kevin Yamauchi <kevin.yamauchi@gmail.com>

brisvag committed:
fix typo

brisvag committed:
eextract and fix logic for data_slice->indices/slices conversion, and add test for it

brisvag committed:
add some comments, streamline the make_full function

brisvag committed:
add connection to current step with big comment explaining

brisvag committed:
fix benchmark

brisvag committed:
Merge branch 'feature/thick-slice-layers' of github.com:brisvag/napari; branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
fix benchmark

kevinyamauchi reviewed (approved):
LGTM! Thank you @brisvag !

brisvag committed:
Merge branch 'main' of github.com:napari/napari into feature/thick-slice-layers

brisvag committed:
fixed
